cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -192,11 +202,41 @@ class _CouldNotResolve(ValueError):
 
 class _ComponentResolver:
     def __init__(
-        self, package_mapping: Mapping[Locator, Package], project: Project, output_dir: RootedPath
+        self,
+        package_mapping: Mapping[Locator, Package],
+        patch_locators: list[PatchLocator],
+        project: Project,
+        output_dir: RootedPath,
     ) -> None:
         self._project = project
         self._output_dir = output_dir
         self._package_mapping = package_mapping
+        self._pedigree_mapping = self._get_pedigree_mapping(patch_locators)
+
+    def _get_pedigree_mapping(self, patch_locators: list[PatchLocator]) -> dict[Locator, Pedigree]:
+        """Map locators for dependencies that get patched to their Pedigree."""
+        pedigree_mapping: dict[Locator, Pedigree] = {}
+        for patch_locator in patch_locators:
+            # Filter out builtin patches
```

Comment:
Builtin patches are applied by Yarn itself and tied to a specific version of Yarn. They are used to make certain features of the package manager work correctly: [example](https://yarnpkg.com/getting-started/qa/#why-is-typescript-patched-even-if-i-dont-use-plugnplay).

Do we want to report these?

Reply:
It would make our code appear "more" functional from SBOM perspective because we could report these patches as "cherry-pick" and therefore claim that we have "some" way of detecting the right type for the patch :D. In all seriousness, I think it would be more factually correct if we did, but based on the way they document it I'd say it all depends on how transparent these are to the user, IOW can we see the patch in the lockfile? Seems like we can, so we should report it as such IMO.

Reply:
My first impression is that there's no harm in reporting them, but then, `unnoficial` seems an incorrect way to report them in the sbom, an apparently there's no better term for it in CycloneDx.

Reply:
So, here's the explanation: https://cyclonedx.org/docs/1.5/json/#components_items_pedigree_patches_items_type

I think we can keep `unnoficial` for user-added patches, and maybe `monkey` for official Yarn patches? I can't find any better way to express this, the choices in CycloneDX are very limiting.

Reply:
Just an FYI, for builtin patches we're probably going to need to clone the yarn (berry) repo and figure out a [.ts file](https://github.com/yarnpkg/berry/tree/master/packages/plugin-compat/sources/patches) there to point to rather than a standard diff, since the patches don't exist in the user repo and the URL for them does not appear in yarn info/lock.

Reply:
> So, here's the explanation: https://cyclonedx.org/docs/1.5/json/#components_items_pedigree_patches_items_type
> 
> I think we can keep `unnoficial` for user-added patches, and maybe `monkey` for official Yarn patches? I can't find any better way to express this, the choices in CycloneDX are very limiting.

`monkey` doesn't IMO describe them well - these are official patches that are applied on top of Yarn by Yarn itself and they carry those patches over to newer releases ---> this is a definition of cherry picking.

Reply:
> Just an FYI, for builtin patches we're probably going to need to clone the yarn (berry) repo and figure out a [.ts file](https://github.com/yarnpkg/berry/tree/master/packages/plugin-compat/sources/patches) there to point to rather than a standard diff, since the patches don't exist in the user repo and the URL for them does not appear in yarn info/lock.

That's unfortunate, but if our goal is to accurately record this in the SBOM, I guess we need to something like that. What ref/branch would we be cloning in that case? Always "main"? Or the latest `vX` tag?

I'm also wondering if fetching the official Yarn `vX` tarball, unpacking it and walking the hierarchy to identify the path wouldn't be slightly cheaper AND/OR faster than git operations on blobless shallow git clones.

Reply:
> these are official patches that are applied on top of Yarn by Yarn itself and they carry those patches over to newer releases

They applied by Yarn on top of a few libraries, like Typescript ([e.g.](https://github.com/cachito-testing/cachi2-yarn-berry/blob/regular-installs/yarn.lock#L1331))

Reply:
Why wouldn't builtin patches be "unofficial"? If yarn patches Typescript, it may be an official patch in the context of Yarn, but it almost certainly isn't in the context of Typescript, which is the Component being reported.

Reply:
Well, my reasoning came from wanting to set them apart, but on the other hand, the purls will still carry the `~builtin` qualifier, right? I'm fine to just make them `unofficial`.

Reply:
> Why wouldn't builtin patches be "unofficial"? If yarn patches Typescript, it may be an official patch in the context of Yarn, but it almost certainly isn't in the context of Typescript, which is the Component being reported.

But Typescript is Yarn's own dependency, not user's one, correct? IOW It is Yarn who needs it not user's app, doesn't it? If I'm wrong in this base assumption you can just stop reading here :), no need to continue the discussion.

Otherwise, it is Yarn's own patch applied on top of Yarn code base which doesn't sound unofficial to me.

In any case, I don't dwell too much on whether this will be unofficial or not in the SBOM, but there must be a clear record on why we think it is the case, it cannot sound just like an arbitrary decision.

Reply:
> But Typescript is Yarn's own dependency, not user's one, correct?

Typescript would be a dependency of the user project. The builtin patches are only applied if the dependency they are targeted to patch is included

Reply:
In which case my argument is invalid -> "unofficial" it is, sorry for the noise and thank you for the clarification.